src/vs/vscode.d.ts

```diff
@@ -968,6 +968,17 @@ declare module 'vscode' {
 		edit(callback: (editBuilder: TextEditorEdit) => void, options?: { undoStopBefore: boolean; undoStopAfter: boolean; }): Thenable<boolean>;
 
 		/**
+		 * Inserts the given snippet template and enters snippet mode.
+		 *
+		 * If the editor is already in snippet mode, insertion fails and the returned promise resolves to false.
+		 *
+		 * @param template The snippet template to insert
+		 * @param posOrRange The position or replacement range representing the location of the insertion.
+		 * @return A promise that resolves with a value indicating if the snippet could be inserted.
+		 */
+		insertSnippet(template: string, posOrRange: Position | Range): Thenable<boolean>;
```

Comment:
Agreed that you should be able to call this without being concerned with location. Sharing an argument for both position or range seems a little awkward as well. So long as I can alter the selection before starting. How should this work for multiple selections?

In the case of overloading `TextEditor.edit`, if that involves changes to `TextEditorEdit`, how does entering snippet mode fit in with subsequent insertions, replacements, etc.? It seems like this might have more in common with `CompletionItem` than editing.
 
Another thought: Should we supply a callback to know when snippet mode was exited?